cassiopeia/tools/branch_length_estimator/IIDExponentialBayesian.py

```diff
@@ -0,0 +1,805 @@
+        self,
+        tree: CassiopeiaTree,
+        node: str,
+        mutation_rate: float,
+        birth_rate: float,
+        sampling_probability: float,
+        discretization_level: int,
+        epsrel: float = 0.01,
+    ) -> np.array:
+        """
+        Numerical posterior time inference under the model.
+
+        This method is used for testing the implementation of posterior_time of
+        the model.
+
+        The posterior time distribution of a node, numerically computed, i.e.:
+        P(node time = t | character states, tree topology) for t in [0, T]
+        where T is the discretization_level.
+
+        Note that this method is only fast enough for small trees. It's
+        run time scales exponentially with the number of internal nodes of the
+        tree.
+
+        Args:
+            tree: The CassiopeiaTree containing the tree topology and all
+                character states.
+            node: An internal node of the tree, for which to compute the
+                posterior time distribution.
+            mutation_rate: The mutation rate of the model.
+            birth_rate: The birth rate of the model.
+            sampling_probability: The sampling probability of the model.
+            discretization_level: The number of timesteps used to discretize
+                time. The output thus is a vector of length
+                discretization_level + 1.
+            epsrel: The degree of tolerance for the numerical integrals
+                performed.
+
+        Returns:
+            P(node time = t | character states, tree topology) for t in [0, T]
+            where T is the discretization_level.
+        """
+        numerical_log_joints = self.numerical_log_joints(
+            tree=tree,
+            node=node,
+            mutation_rate=mutation_rate,
+            birth_rate=birth_rate,
+            sampling_probability=sampling_probability,
+            discretization_level=discretization_level,
+            epsrel=epsrel,
+        )
+        numerical_posterior = np.exp(
+            numerical_log_joints - numerical_log_joints.max()
+        )
+        numerical_posterior /= numerical_posterior.sum()
+        return numerical_posterior.copy()
+
+    def _simple_inference_sanity_check(self):
+        """
+        Tests that the likelihood computed from each leaf node is correct.
+        """
+        tree = self._tree
+        for leaf in tree.leaves:
+            model_log_likelihood_up = (
+                self._up(
+                    leaf,
+                    self._discretization_level,
+                    _get_number_of_mutated_characters_in_node(tree, leaf),
+                )
+                - np.log(self._birth_rate * 1.0 / self._discretization_level)
+                + np.log(self._sampling_probability)
+            )
+            np.testing.assert_approx_equal(
+                self._log_likelihood, model_log_likelihood_up, significant=2
+            )
```

Comment:
AFAIK, it's not good practice for tests to access private members, because changing an implentation detail causes the tests to break. What one can do is therefore to refactor these tests that involve private members (like `_up` in this case) to a new method (`_simple_inference_sanity_check` in this case). Sure, the testing method it's private, but at least now the tests only call one private member: `_simple_inference_sanity_check`, and if implementation details change the testing code doesn't have to be changed too.

Reply:
I removed this test completely, and have decided to only test the public API.